### PR TITLE
feat: Add `EnvironmentsByKubernetes` api query

### LINKED ISSUE #2684: Add API query to list all projects on a cluster
- New `projectsByKubernetes` query
- Filter by `id`, `createdAfter`
- Sort by `name`, `created`

### PR DESCRIPTION
**Pull request by rocketeerbkw**
# Checklist

- [x] Affected Issues have been mentioned in the Closing issues section
- [ ] Documentation has been written/updated
- [x] PR title is ready for changelog and subsystem label(s) applied

# Closing issues

closes #2684

### CODE REVIEW

**Comment by rocketeerbkw** [comment]
This query can be executed by admins and regular users. For regular users we have to loop over every project and do a keycloak check to ensure they can view the openshift data (by default, only project maintainers can view it).

This will be really slow if anyone has a lot of projects and put more strain on keycloak, so I'm wondering if we should even allow this for non-admins. The upside is we're not using this in the UI or the CLI, but do we wanna let those horses out of the barn right now?

**Comment by rocketeerbkw** [comment]
I only added `projectsByKubernetes` and didn't add `projectsByOpenshift` because we want to get rid of openshift references eventually anyway.

**Comment by tobybellwood** [comment]
This makes sense, but a pivot to EnvironmentsByKubernetes would account for prod/dev split DeployTargets.

@CGoodwin90 to refactor

File: services/api/src/typeDefs.js
Comment (shreddedbacon) [inline services/api/src/typeDefs.js]: The `allEnvironments` query has a filter for the `type` too. I think it could be useful to add this filter to this resolver too. Then you could do all production environments on cluster X

File: services/api/src/resources/environment/resolvers.ts
Comment (shreddedbacon) [inline services/api/src/resources/environment/resolvers.ts]: You only need to do the `openshift:viewAll` permission check here, but that means the query will only work for legacy tokens until #3437 lands. But this is the same behaviour that `allEnvironments` query has with `environment:viewAll` which is also resolved by the same PR.

**Review by shreddedbacon — CHANGES_REQUESTED** [review]

File: services/api/src/resources/environment/resolvers.ts
Comment (shreddedbacon) [inline services/api/src/resources/environment/resolvers.ts]: For non-admin users using this query, it fails because the way this works was changed in the keycloak optimisations PR.

You'll need to update this to use the following
```
const userProjectRoles = await models.UserModel.getAllProjectsIdsForUser({
  id: keycloakGrant.access_token.content.sub,
}, keycloakUsersGroups);
userProjectIds = getUserProjectIdsFromRoleProjectIds(userProjectRoles);
```

You'll also need to update the resolver function to include `keycloakUsersGroups` and also import 
```
import { getUserProjectIdsFromRoleProjectIds } from '../../util/auth';
```

**Review by shreddedbacon — CHANGES_REQUESTED** [review]

File: services/api/src/resources/openshift/helpers.ts
Comment (shreddedbacon) [inline services/api/src/resources/openshift/helpers.ts]: Throw the same `invalid input for kubernetes` error here

**Review by shreddedbacon — CHANGES_REQUESTED** [review]
One last thing, but otherwise all good!

**Review by shreddedbacon — APPROVED** [review]

**Pull request merged**
